### PR TITLE
Fix workspaces filepath issue

### PR DESCRIPTION
**Pull request by ckundo**
## Summary

This change fixes an issue with projects using yarn workspaces failing to report filepath correctly.

## Details

- Add configuration option for workspace package root:
```
  let app = new EmberApp(defaults, {
    babel: {
      plugins: [
        [
          require.resolve('babel-plugin-ember-test-metadata'),
          {
            enabled: true,
            workspaceRoot: 'packages',
            packageName: defaults.project.pkg.name,
          }
        ]
      ],
    }
  });
```

## Testing

- Add workspace app called `extended` and corresponding scenario tests.

### CODE REVIEW

**Comment by scalvert** [comment]
Hey @ckundo, thanks for taking another stab at this.

Before we go too much further, could we clearly articulate the precise scenario we're needing to support here? As it stands, we support:

- `classic`
- `classic` with `in-repo` tests (via `ember-add-in-repo-tests`)
- `embroider`
- `embroider` with `in-repo` tests (via `ember-add-in-repo-tests`)
- scenarios where the app is in an unconventional root, such as in a subdirectory, via the `packageRoot` configuration (which handles cases where the subdirectory name matches the app's name, [causing a doubling of the name in the path](https://github.com/babel-plugin-ember-test-metadata/babel-plugin-ember-test-metadata/blob/4ed5163920c5a0c631f512a125d5a366892a2edd/packages/babel-plugin-ember-test-metadata/__tests__/get-normalized-file-path-test.js#L7-L20))

If you could provide the pre-normalized file path (the one that the babel plugin receives), and the expected post-normalized file path (after we've adjusted it via the plugin), that'd be really useful in understanding what we're targeting here.

**Review by thousand — COMMENTED** [review]
A few comments here and there. Overall, I can't really comment on the validity of the approach since i'm not an SME in babel/ember-cli path stuff, so +1 to @scalvert's request for more clarity on how this fixes the underlying issues.

File: packages/test-scenarios/__tests__/__fixtures__/with-packages-test.js [hunk 1]
```diff
@@ -0,0 +1,12 @@
+import { module, test } from 'qunit';
+import { getTestMetadata } from '@ember/test-helpers';
+
+module('Acceptance | with-packages-test', function (hooks) {
+  hooks.beforeEach(function () {
+    // noop
+  });
```
[the remark this answers is not in the document]
Comment (scalvert) [reply packages/test-scenarios/__tests__/__fixtures__/with-packages-test.js:7]: This plugin actually tests adding code to the test files, so we want to ensure that given a specific file with a specific structure, we're transforming that file into the correct output structure. I _believe_ this `beforeEach` call is part of that test.

**Comment by ckundo** [comment]
@scalvert sure thing!

Here's what getting passed to the babel plugin. File path on disk for this example is: `/path/to/workspace/example-web/packages/extended/tests/acceptance/example-test.js`, and the name of the workspace package is `extended`.

```
filename: "/path/to/workspace/example-web/packages/extended/extended/tests/acceptance/example-test.js"
projectName: "example-web"
root: "/path/to/workspace/example-web/packages/extended"
```

we expect the post-normalized path to be `packages/extended/tests/acceptance/example-test.js` but instead it is `packages/extended/extended/tests/acceptance/example-test.js`

it looks to me like ember-cli or else broccoli is may be prepending the package name to the relative filepath before adding it to the root path.

**Comment by scalvert** [comment]
Right, this was known to us when originally developing the plugin. This is how the builds work - broccoli will append the package name for the app into the real path to the app, which includes the app's directory. This is the original reason we devised adding the `packageName` option; it indicated a split point for us to trim off the rest of the path at the right segment. You can see this in [the original tests](https://github.com/babel-plugin-ember-test-metadata/babel-plugin-ember-test-metadata/blob/master/packages/babel-plugin-ember-test-metadata/__tests__/get-normalized-file-path-test.js#L9). the first `classic` segment in the path is the real directory name on disc, and the second `classic` segment is the app name from the app's package.json concatenated on the end.

I think you just need to change the `packageName` option in plugin's config in the project's `ember-cli-build.js` file to resolve this. I've demonstrated this in [two tests in this branch](https://github.com/babel-plugin-ember-test-metadata/babel-plugin-ember-test-metadata/commit/0727c8db80c1e1e0da01173c576c48145771ac22), which both pass.

Am I missing something?

**Comment by ckundo** [comment]
@scalvert totally, that would work in the most common scenario where there was a separate babel config for each package. We're dealing with a scenario where there's a single, shared babel build config across packages. I'm not sure how common that is in ember projects, but if we want I can update the scenario test with multiple packages sharing a config.

**Comment by scalvert** [comment]
Yep, I was aware that the app in question uses a single, shared babel config. My understanding is that this config issue manifested when the repo was converted from a "conventional" structure to a monorepo via yarn workspaces. If that's correct, the app already used a centralized config, so how is it different now if it's simply residing in a different project location? Also, since the repo in question is under our control (being cryptic here for other readers, since it's an internal repository to our company), we can likely make changes to that config, eg. split into a few different configs that work for the configurations we need.

As for the proposed change, it adds a new config property to specify the path segment to target, which is what the original `packageName` config option did. How is the new setting different, and how is it going to address your comment above differently?

**Comment by thousand** [comment]
@ckundo i think it'd be good to rule out the config problem in the app regardless of the structure. You're right that there's some custom infra involved, but a lot of that is aiming to create common or centralized configs to reduce the amount of boilerplate needed in setting up those packages, and from what @scalvert is saying here it seems likely it's in that stuff rather than here.

**Comment by scalvert** [comment]
Just to add to what @thousand commented, my preference would be to drive these custom app structures via configuration so the uniqueness can be declaratively supported rather than adding complexity to this plugin for particular use cases. This is a more extensible approach, and moves the specific config complexity to the consumer, not the underlying babel plugin.

**Comment by ckundo** [comment]
@scalvert @thousand thanks for the feedback, closing in favor of fixing the configuration on the consumer side.

**Pull request closed without merging**